src/builtins.c

```diff
@@ -549,6 +550,32 @@ JL_CALLABLE(jl_f__apply)
     return result;
 }
 
+// this is like `_apply`, but with quasi-exact checks to make sure it is pure
+JL_CALLABLE(jl_f__apply_pure)
```

Comment:
You have to expand the diff for inference.jl (and even then, my brower's ctrl-f doesn't seem to like finding it). The function semantically is invalid, so I don't encourage using it, but it's a necessary hack for now.